File: analytics/correlation.py
"""Moving-average sentiment and lead-lag correlation vs price."""

from __future__ import annotations

import math
from statistics import mean


def moving_average(values: list[float], window: int) -> float:
    if not values:
        return 0.0
    slice_ = values[-window:] if window > 0 else values
    return float(mean(slice_))


def pearson(xs: list[float], ys: list[float]) -> float:
    n = min(len(xs), len(ys))
    if n < 4:
        return 0.0
    xs, ys = xs[:n], ys[:n]
    mx, my = mean(xs), mean(ys)
    cov = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    vx = sum((x - mx) ** 2 for x in xs)
    vy = sum((y - my) ** 2 for y in ys)
    den = math.sqrt(vx * vy)
    if den < 1e-12:
        return 0.0
    return max(-1.0, min(1.0, cov / den))


def returns(prices: list[float]) -> list[float]:
    out = []
    for prev, cur in zip(prices, prices[1:]):
        out.append(0.0 if prev <= 0 else (cur - prev) / prev)
    return out


def lead_lag(prices: list[float], sentiment: list[float], bar_minutes: int = 15) -> dict:
    r_price = returns(prices)
    r_sent = sentiment[1 : 1 + len(r_price)]
    n = min(len(r_price), len(r_sent))
    a, b = r_price[:n], r_sent[:n]
    best_lag, best_corr = 0, pearson(a, b)
    max_lag = min(8, n // 5)
    for lag in range(-max_lag, max_lag + 1):
        if lag == 0:
            continue
        if lag > 0:
            xs, ys = b[: n - lag], a[lag:]
        else:
            k = -lag
            xs, ys = b[k:], a[: n - k]
        c = pearson(xs, ys)
        if abs(c) > abs(best_corr):
            best_corr, best_lag = c, lag
    return {
        "pearson": best_corr,
        "lag_minutes": best_lag * bar_minutes,
    }

File: src/lib/sentra/correlation.ts
import type { SeriesPoint } from "./types";

function pearson(xs: number[], ys: number[]): number {
  const n = Math.min(xs.length, ys.length);
  if (n < 4) return 0;
  let sx = 0;
  let sy = 0;
  let sxx = 0;
  let syy = 0;
  let sxy = 0;
  for (let i = 0; i < n; i += 1) {
    const x = xs[i] ?? 0;
    const y = ys[i] ?? 0;
    sx += x;
    sy += y;
    sxx += x * x;
    syy += y * y;
    sxy += x * y;
  }
  const cov = sxy - (sx * sy) / n;
  const vx = sxx - (sx * sx) / n;
  const vy = syy - (sy * sy) / n;
  const den = Math.sqrt(vx * vy);
  if (den < 1e-12) return 0;
  return Math.max(-1, Math.min(1, cov / den));
}

function returns(values: number[]): number[] {
  const out: number[] = [];
  for (let i = 1; i < values.length; i += 1) {
    const prev = values[i - 1] ?? 0;
    const cur = values[i] ?? 0;
    if (prev <= 0) {
      out.push(0);
      continue;
    }
    out.push((cur - prev) / prev);
  }
  return out;
}

export function leadLag(series: SeriesPoint[], barMinutes: number) {
  const prices = series.map((p) => p.price);
  const sents = series.map((p) => p.sentiment);
  const rPrice = returns(prices);
  const rSent = sents.slice(1);
  const n = Math.min(rPrice.length, rSent.length);
  const a = rPrice.slice(0, n);
  const b = rSent.slice(0, n);

  const maxLag = Math.min(8, Math.floor(n / 5));
  let bestLag = 0;
  let bestCorr = pearson(a, b);

  for (let lag = -maxLag; lag <= maxLag; lag += 1) {
    if (lag === 0) continue;
    let xs: number[];
    let ys: number[];
    if (lag > 0) {
      xs = b.slice(0, n - lag);
      ys = a.slice(lag);
    } else {
      const k = -lag;
      xs = b.slice(k);
      ys = a.slice(0, n - k);
    }
    const c = pearson(xs, ys);
    if (Math.abs(c) > Math.abs(bestCorr)) {
      bestCorr = c;
      bestLag = lag;
    }
  }

  const lagMinutes = bestLag * barMinutes;
  let interpretation: string;
  if (Math.abs(bestCorr) < 0.12) {
    interpretation =
      "No reliable link this window — sentiment and price are moving independently.";
  } else if (lagMinutes < 0) {
    interpretation = `Sentiment leads price by ${Math.abs(lagMinutes)}m (r=${bestCorr.toFixed(2)}). Voice is arriving before the print.`;
  } else if (lagMinutes > 0) {
    interpretation = `Price leads sentiment by ${lagMinutes}m (r=${bestCorr.toFixed(2)}). The tape is setting the conversation.`;
  } else {
    interpretation = `Contemporaneous move (r=${bestCorr.toFixed(2)}). Sentiment and price are printing together.`;
  }

  return {
    pearson: bestCorr,
    lagMinutes,
    interpretation,
  };
}
